File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows, cols = len(grid), len(grid[0])
        visit = set()
        iSize = []
        iCnt = 0
        def bfs(r, c):
            s = 1
            q = [(r, c)]
            visit.add((r, c))
            direction = [(1, 0), (-1, 0), (0, 1), (0, -1)]
            while q:
                r1, c1 = q.pop(0)
                for x, y in direction:
                    r2, c2 = r1 + x, c1 + y
                    if r2 in range(rows) and c2 in range(cols) and grid[r2][c2] == '1' and (r2, c2) not in visit:
                        q.append((r2, c2))
                        visit.add((r2, c2))
                        s += 1
            iSize.append(s)

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '1' and (r, c) not in visit:
                    bfs(r, c)
                    iCnt += 1
        print(len(visit)) # 1 개수
        print(iSize) # 섬의 사이즈
        print(iCnt) # 섬 개수
        return iCnt
        

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        if (grid.length == 0 || grid[0].length == 0) return 0;
        int count = 0;
        for(int r = 0; r < grid.length; r++) {
            for(int c = 0;  c < grid[0].length; c++) {
                if(grid[r][c] == '1') {
                    count++;
                    bfs(grid, r, c);
                }
            }
        }
        return count;
    }
    public void bfs(char[][] grid, int r, int c) {
        if(r < 0 || c < 0 || r >= grid.length || c >= grid[0].length || grid[r][c] == '0') return;
        grid[r][c] = '0';
        bfs(grid, r+1, c);
        bfs(grid, r-1, c);
        bfs(grid, r, c+1);
        bfs(grid, r, c-1);
        
    }
}